feat: Allow FQCN as fixer name and configuration key for rules

Wirone opened the pull request:
Fixes #7062

- Allow FQCN as fixer name
- Allow lowercased vendor in custom fixers (foo/bar)
- Allow using FQCN in rulesets configuration (as a array key, for including/configuring/excluding rule in the set)

Wirone committed:
Improve fixer names' validator

- Allow FQCN as fixer name
- Allow lowercased vendor in custom fixers (foo/bar)

SpacePossum commented:
* Allow FQCN as fixer name

maybe do not allow custom rule name start with `PhpCsFixer\` ? Than we "reserve" the namespace for build in rules so we can later rename or alias the build in rules with the FQCN without possible conflicts with custom rules.

Wirone committed:
Disallow FQCN rule names starting with `PhpCsFixer\`

kubawerlos reviewed (dismissed):
Technically it's fine, but I'd love to hear from @keradus about this change as the naming come from ancient times of `fabpot/php-cs-fixer` that only he remembers :)

keradus commented:
sorry, not sure if I pick up the question around fabpot/* namespace, can you clarify?

Wirone commented:
@keradus I believe @kubawerlos just wonders if you're in favour of such change or maybe there were some ancient decisions about naming convention that you wouldn't like to change 😉.

mvorisek commented on src/FixerNameValidator.php:
```diff
@@ -27,6 +27,7 @@ public function isValid(string $name, bool $isCustom): bool
 
-        return 1 === Preg::match('/^[A-Z][a-zA-Z0-9]*\/[a-z][a-z0-9_]*$/', $name);
+        // See: https://regex101.com/r/UngJUn/4
+        return 1 === Preg::match('/^[a-zA-Z][a-zA-Z0-9]*(?!$)((\/[a-z][a-z0-9_]*)?|((?i)(?<!PhpCsFixer)(?-i)\\\\[a-zA-Z][a-zA-Z0-9_]*)*){1}$/', $name);
```
the regex seems way too complex

`a-zA-Z` and `(?i)` can be replaced with `i` global modifier

`{1}` is useless, one repetition is by default ;-)

Wirone replied:
[v4](https://regex101.com/r/UngJUn/4): 6 matches (1 002 steps, 0.5ms)
[v5](https://regex101.com/r/UngJUn/5): 7 matches (827 steps, 0.1ms)

Quicker, but `i` modifier allows `VENDOR/SHOUT` which previously (in my regex, but also in original one) was rejected (part after `/`).

mvorisek replied:
FQCN should validate againts `TypeExpression::REGEX_IDENTIFIER` regex introduced in #7010

Wirone replied:
I don't know if it's required to check for class existence at this point - we only need to know if it's `vendor/name` (current custom rule naming convention) or FQCN-like string. If rule's name used in the config (like `Foo::class => true`) is invalid, then it will fail anyway.

keradus replied:
btw, `[Foo::getName() => true]`

keradus replied:
I do not see big benefit of using `Foo::class` over `Foo::getName()`

Wirone replied:
@keradus `FixerInterface::getName()` is not static, so it should be `(new FooFixer())->getName()` which does not make sense.

keradus commented on tests/FixerNameValidatorTest.php:
```diff
@@ -50,15 +50,25 @@ public static function provideIsValidCases(): array
             ['4vendor/foo', true, false],
             ['Vendor/foo', true, true],
             ['FooBar/foo', true, true],
+            ['foo\\bar', true, true],
```
original idea was to support format of `vendor/rule`. now you also add `vendor\rule`.
not sure if I would like to keep 2, `/` and `\`, alternative syntaxes
I would like to keep explicit `vendor/` part - and then whatever after

Wirone replied:
@keradus did you read #7062 as a background for this change? This is whole point of this PR - to allow FQCNs as rule name. Right now it's impossible to use rules from `symplify/coding-standard` and there are several other drawbacks of using "random" strings.

keradus replied:
if we expose our integration point with interface and we expect specific format, ie `vendor/rule`, and some external library decides to not use that format, I would rather expect that format that is expected is provided, rather than we change expectations.
^^ that's about the principle approach in general.

### point 1
for this case specifically, i would love to see have single way how we format of rules. and if it's sometimes `foo/bar`, and sometimes `foo\bar`, then it's off and I would love to avoid it.

### point 2
if we consider to touch possible format of name for custom rules, i would suggest to combine the alignment with possible format of name for custom rulesets

### point 3
if using random string is a drawback, let's solve that drawback properly - change of string format is not solving them

Wirone replied:
> if we expose our integration point with interface and we expect specific format, ie vendor/rule, and some external library decides to not use that format, I would rather expect that format that is expected is provided, rather than we change expectations.

This PR is not created in order to align with userland convention, but to address several concerns raised in #7062. So this is not about changing expectation, rather about changing general approach. Using FQCNs enables several advantages - code navigation, usage detection, visual deprecation indicator, easy refactoring etc.

In terms of other points:

1. I think this can be only for transitional period, because I think FQCN would be better as a contract in the future. At this point we can allow both and systematically, step by step, shift in FQCN direction. `vendor/name` is kept for backward compatibility (allows only 1 `/something` segment), while newly introduced `vendor\name` allow 1+ `\something` segment(s) and is forward-compatible. 
2. Yes, we could allow FQCNs as a name of custom rule sets too.
3. Not sure what you mean. We can't solve it just like that, as I wrote in the point 1 - it requires transition period. Allowing FQCNs is first step of the process which may lead to full switch to FQCNs in v4.

PS. Support for FQCNs for built-in fixers can be done with 2 lines, by adding `$this->fixersByName[get_class($fixer)] = $fixer;` to `FixerFactory::registerFixer()` and aligning `$availableFixers` in `ConfigurationResolver::validateRules()`. I could even add it in this PR so people can instantly start configuring Fixer using FQCNs, at least when it comes to fixers, not rulesets.

Wirone replied:
Built-in fixers are now registered under FQCN too, so these can be referenced by FQCN (as `PhpCsFixerSet` shows).

keradus reviewed (dismissed):
sorry, but for now I'm not convinced to use FQCN to setup fixers.

ostrolucky commented:
Why can't we allow it at least, though? I would also welcome it, to make it compatible with 3rd party fixers.

Wirone commented:
@keradus I'be been using FQCN for years in ECS and I find it better because of reasons mentioned in the #7062 and here. It simplifies so much, you can easily check if provided FQCN refers to class that implements required contract, no need for registration flow etc. I don't get why don't see those advantages...

Wirone committed:
Simplify and speed up regex

Wirone committed:
Support for FQCNs in rules configuration

It's fully backward compatible, but allows using FQCNs as a keys in rules' config, instead of short names. This enables code navigation, usage discoverability etc.

 Conflicts:
	src/Documentation/RuleSetDocumentationGenerator.php
	src/RuleSet/Sets/PhpCsFixerSet.php

acairns commented:
I support the idea of using FQCNs for ruleset configuration and as a rule name (like ECS' rules do and that's why they can't be used right now in Fixer)

Nice work @Wirone 👏

keradus reviewed (changes requested):
marking as blocked to prevent too-eagerly-fast merge.
I want to think over this first and ensure it's approached in direction we want to go long-run. Eg:
- we should have same approach for rule and for ruleset
- one of long term idea was to make config code-less and have static file (eg JSON). This direction is going to support config as code, that I would rather sunset for long-run

I treat it more like RFC for now, didn't check the implementation

Wirone commented:
>  we should have same approach for rule and for ruleset

It was already discussed - I agree, but it won't happen in this PR. It can be added later as a part of a longer plan (just like custom rulesets).

keradus commented:
I do not recall any alignment on direction for those elements I mentioned. Maybe because tight year and lack of capacity as high as yours.

Wirone commented:
Take a look [here](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/7064#discussion_r1235960030) (and your comment above), I already addressed your concerns about long-term vision and consistency between rules and rulesets.

keradus commented:
I understand that you have the vision and that we didn't had a chance to get enough attention from my side to conclude on it - and while you have idea, I have no conclusion. For me it was definitely 8.3. It's done for few days, time to recover and then prioritise topics to focus on further (opening more till we conclude on previous not helping me).

gharlan commented:
I think it is a really good move! I would like to use FQCNs for built-in rules, rule-sets and custom rules. The old names should be marked as deprecated in my opinion. Therefore, in the long term, it should only be possible to configure the rules via the FQCNs.

Wirone committed:
Fix PHPStan issues

 Conflicts:
	dev-tools/phpstan/baseline/offsetAccess.notFound.php

gharlan commented:
In my rector config I like to use aliases for partial imports. Example:

```php
use Rector\CodeQuality\Rector as CodeQuality;
use Rector\CodingStyle\Rector as CodingStyle;
use Rector\Config\RectorConfig;
use Rector\Php80\Rector as Php80;
use Rector\Php83\Rector as Php83;
use Rector\Symfony\Symfony62\Rector as Symfony62;

return RectorConfig::configure()
    // ...
    ->withRules([
        CodeQuality\FuncCall\SingleInArrayToCompareRector::class,
        CodeQuality\NullsafeMethodCall\CleanupUnneededNullsafeOperatorRector::class,
        CodingStyle\ClassConst\RemoveFinalFromConstRector::class,
        Php80\Class_\ClassPropertyAssignToConstructorPromotionRector::class,
        Php83\ClassMethod\AddOverrideAttributeToOverriddenMethodsRector::class,
        Symfony62\MethodCall\SimplifyFormRenderingRector::class,
    ])
;
```

I would like to do something similar in my php-cs-fixer config:

```php
use PhpCsFixer\Config;
use PhpCsFixer\Fixer\Comment;
use PhpCsFixer\Fixer\FunctionNotation;
use PhpCsFixer\Fixer\Operator;
use PhpCsFixer\Fixer\StringNotation;
use PhpCsFixer\RuleSet\Sets;
use PhpCsFixerCustomFixers\Fixer as Kubawerlos;

return new Config()
    ->setRules([
        Sets\SymfonySet::class => true,
        Sets\SymfonyRiskySet::class => true,
        Sets\PHP84MigrationSet::class => true,
        Sets\PHP82MigrationRiskySet::class => true,

        Comment\MultilineCommentOpeningClosingFixer::class => true,
        FunctionNotation\MultilinePromotedPropertiesFixer::class => true,
        FunctionNotation\SingleLineThrowFixer::class => false,
        FunctionNotation\UseArrowFunctionsFixer::class => false,
        Operator\ConcatSpaceFixer::class => ['spacing' => 'one'],
        StringNotation\HeredocToNowdocFixer::class => true,

        Kubawerlos\PhpdocSingleLineVarFixer::class,
    ])
;
```

ruudk commented:
It's now 2 years later. What do you think about this @keradus? The DX could be so much better if we could easily click through the code of the actual fixer, instead of taking the snake case name, convert it to pascal case and then search it either in GitHub or the IDE. Also building and using custom fixers becomes a breeze.

Rector really nailed this and it's also why Easy Coding Standard works with FQCN's.

Wirone commented:
@ruudk (and others interested in this PR), FYI: @keradus gave a green light for this approach, so FQCN support *is going to be added*, _BUT_ we need to align from technical perspective, so this PR enters regular code review phase and will be merged after approvals are given 🥳. Since it's a huge step (even though it's not technically that complex), @keradus wants to validate it properly to not introduce bad API for the end users. Please be patient, we'll process it in our own pace. I already rebased the branch so it's in sync with latest `master` and can be reviewed.

Wirone committed:
Integration test showing usage of FQCNs as fixer names in ruleset config

Wirone committed:
Normalise rulesets after validation

Wirone committed:
More unit tests for normalising rulesets

# Conflicts:
#	tests/RuleSet/RuleSetTest.php

kubawerlos commented on tests/RuleSet/RuleSetTest.php:
```diff
@@ -168,59 +162,125 @@ public function testResolveRulesWithMissingRuleValue(): void
+            ],
+        ];
+
+        yield 'same rule configure three times with different approaches for names, but with leading backslash in raw string\'s FQCN' => [
+            [
+                '\PhpCsFixer\Fixer\ArrayNotation\ArraySyntaxFixer' => true,
+                'array_syntax' => false,
+                ArraySyntaxFixer::class => true,
+            ],
+            [
+                'array_syntax' => true,
+            ],
+        ];
```
```suggestion
        yield 'same rule configure three times with different approaches for names, but with leading backslash in raw string\'s FQCN' => [
            [
                '\PhpCsFixer\Fixer\ArrayNotation\ArraySyntaxFixer' => true,
                'array_syntax' => ['syntax' => 'long'],
                ArraySyntaxFixer::class => ['syntax' => 'short'],
            ],
            [
                'array_syntax' => ['syntax' => 'short'],
            ],
        ];
```
There are 2 `true` config values, and it is not clear which one ends in the config. Let's make it clear.

Wirone committed:
Use different values for all 3 approaches of configuring the same rule

Co-authored-by: Kuba Werłos <werlos@gmail.com>

Wirone commented on src/RuleSet/RuleSet.php:
```diff
@@ -62,9 +63,45 @@ public function __construct(array $set = [])
+     *
+     * @return array<string, array<string, mixed>|bool>
+     */
+    public static function normaliseConfig(array $rulesConfig): array
```
I've checked and American version is used multiple times in the codebase, while British only in my changes. I'll update it.

Wirone committed:
Simplify checking whether FQCN references FixerInterface

Co-authored-by: Gregor Harlan <mail@gh01.de>

Wirone committed:
Improve `RuleSet::normalizeRuleName()`

keradus commented on dev-tools/phpstan/baseline/offsetAccess.notFound.php:
```diff
@@ -318,7 +318,7 @@
 ];
```
...about https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/7064#issuecomment-3233590502 
(i would love it to be a thread in first place)

Wirone committed:
Apply fixes after rebase with conflicts resolution

Wirone commented:
It was "impossible" to match these 2 changes fully until now. In #6083 I couldn't add FQCN support because this one was not accepted and merged, here I couldn't cover custom rulesets because it wasn't merged too 🙂. I could, however, add support for rulesets in general, and most probably custom rulesets would work just like that. I don't remember now why I focused only on rules, maybe I overlooked it, or maybe I wanted small steps. Anyway, I will look at it and maybe we can introduce full FQCN support in one go.

keradus commented on src/FixerFactory.php:
```diff
@@ -192,6 +206,7 @@ public function useRuleSet(RuleSetInterface $ruleSet): self
             throw new \UnexpectedValueException($this->generateConflictMessage($fixerConflicts));
         }
 
+        ksort($fixersByName);
```
why we need to sort it ?

Wirone replied:
I guess there is no hard requirement for this, as tests pass without this sorting, but I guess I wanted to have DX-friendly list of rules for easier debugging etc.

keradus reviewed (dismissed):
we shall update docs to state both possibilities

keradus commented on src/RuleSet/RuleSet.php:
```diff
@@ -63,9 +64,40 @@ public function __construct(array $set = [])
+    public static function normalizeRuleName(string $name): string
+    {
+        return is_a($name, FixerInterface::class, true)
+            ? (new $name())->getName()
```
if rule was provided in config as FQCN, did we run FixerNameValidator on it's name?
also, is it in tests somewhere ?

Wirone replied:
Sorry for using AI for this, but it's just faster than refreshing memory after 3 years 🤷.

> `FixerNameValidator` is never called on the FQCN string from config — and that's correct.
>
> Here's why. `FixerNameValidator` lives exclusively in `FixerFactory::registerFixer()` (line 152), which runs at fixer registration time, not at configuration time. It validates the short name a fixer class reports via `getName()`.
>
> When a user provides a FQCN in their config, the sequence in `ConfigurationResolver` is:
> 
> `parseRules()`
>     → raw FQCN key in the rules array
> `validateRules() `
>     `→ new RuleSet([FQCN => true])`
>     `→ normaliseConfig()` converts FQCN → short name
>     → short name checked against `getRegisteredFixerNames()`
> `new RuleSet($rules)`
>     `→ normaliseConfig()` again, short name stored in `$this->rules`
>
> The FQCN is normalized before it touches anything that does name-format validation. The resulting short name was already verified against `FixerNameValidator` when the fixer was originally registered. Unknown or invalid FQCNs that fail to normalize to a known short name are caught by the `$unknownFixers` check in `validateRules()` (line 796–798).
> 
> On the test question: The integration path is covered by `ConfigTest::testConfigWithFullyQualifiedClassNamesAsFixersNames()` (added in this PR), which runs the full fix command with a FQCN-based config and asserts the right fixers were applied. The `FixerNameValidatorTest` additions in the diff (foo\bar, Foo\Bar, etc.) test the regex for a different scenario — custom fixers whose `getName()` itself returns a backslash-style name — not FQCN config keys.
>
> So @keradus' concern is valid to raise but the answer is: the FQCN never needs to pass `FixerNameValidator` because it doesn't survive normalization as a name.

keradus commented on src/Console/ConfigurationResolver.php:
```diff
@@ -842,13 +840,13 @@ private function validateRules(array $rules): void
         }
 
-        foreach ($fixers as $fixer) {
+        foreach ($this->createFixerFactory()->getFixers() as $fixer) {
```
```suggestion
        foreach ($availableFixers as $fixer) {
```

Wirone replied:
This leads to `Error: Call to a member function getName() on string` (it's `list<string>` vs `list<FixerInterface>`).

keradus commented on src/RuleSet/RuleSet.php:
```diff
@@ -63,9 +64,40 @@ public function __construct(array $set = [])
             }
         }
 
+        $set = self::normalizeConfig($set);
```
maybe this can come handy @Wirone ?
https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/9155/commits/0c453877528a62fed969dba565de0488564f1053#diff-67f3a46df79cf522a063924cd6be692b7fe80d147b96b817824694e79ca404a0R179

Wirone replied:
If by "nested config" you mean referencing sets, then `RuleSet::resolveSubset()` handles normalisation too, so every nested set is normalised.

Wirone commented on src/Documentation/RuleSetDocumentationGenerator.php:
```diff
@@ -103,7 +103,7 @@ public function generateRuleSetsDocumentation(RuleSetDescriptionInterface $defin
         }
 
-        $rules = RuleSet::normaliseConfig($definition instanceof AutomaticRuleSetDescriptionInterface
+        $rules = RuleSet::normalizeConfig($definition instanceof AutomaticRuleSetDescriptionInterface
```
Said Kerad in American English 😂. I dropped "Use American english" commit during rebase.

BTW. There are a lot of "normalize"-like usages across codebase.

Wirone commented on src/RuleSet/Sets/PhpCsFixerSet.php:
```diff
@@ -25,10 +27,10 @@ final class PhpCsFixerSet extends AbstractRuleSetDescription
     public function getRules(): array
     {
-        return [
+        return RuleSet::normalizeConfig([
```
Otherwise it fails on `AbstractSetTestCase::assertRules()`. Honestly, I don't remember the exact reason at this point, it's been 3 years since I created this PR 😭.

<img width="1411" height="1011" alt="Image" src="https://github.com/user-attachments/assets/5097bbcf-2ebf-4400-8fa3-5da980fb50c5" />

Wirone committed:
Document rules-by-FQCN usage

Wirone commented on src/RuleSet/RuleSet.php:
```diff
@@ -63,9 +64,40 @@ public function __construct(array $set = [])
+     * and then fine-tune some rules or disable them. Without this normalisation it would be required to always
+     * use only exact names in the config (FQCN only where FQCN is used as name).
+     */
+    public static function normalizeRuleName(string $name): string
```
Unfortunately at this point I don't remember why it was done like this. But I had pretty good project knowledge at the point when I implemented this, and I am pretty sure I did it like this for a reason 😃. I used AI by prompting:

> Try to determine why normalisation was done in `RuleSet` - I am pretty sure I had valid reasons for that considering available architecture. Most probably it is related to how CLI command builds rules from config/CLI options and how resolvers work. I just don't remember at this point, and I don't want to analyse this now.

and here's the summary:

> Why normalization belongs in RuleSet
>
> `RuleSet::getRules()` is a public contract consumed by at least 4 components outside `FixerFactory`, all of which depend on getting back short, normalized names:
>
> | Consumer | File | Why it breaks without normalization in `RuleSet` |
> |---|---|---|
> | **Cache signature** | `src/Cache/Signature.php:94` | Compares rule arrays for cache validity — FQCN vs short name would always be a mismatch, busting the cache on every run |
> | **Describe command** | `src/Console/Command/DescribeCommand.php:750` | `new RuleSet($value)->getRules()` for the "expand" feature — display logic assumes short names |
> | **Fixer doc generator** | `src/Documentation/FixerDocumentGenerator.php:308-309` | `$ruleSet->hasRule($ruleName)` — lookup by short name would not find FQCN-keyed entries |
> | **`ConfigurationResolver::getRules()`** | `src/Console/ConfigurationResolver.php:496` | Returns `$this->getRuleSet()->getRules()` directly to callers including the cache manager |
>
> The complete data flow is:
>
> ```
> Raw user config (may contain FQCNs)
>  → ConfigurationResolver::getRuleSet() → new RuleSet($rules)
>      → normaliseConfig()   ← normalization happens here
>      → resolveSet()
>  → RuleSet::getRules()     ← returns short names, consumed by everyone
>      ├─ FixerFactory::useRuleSet()
>      ├─ Cache::Signature (cache key)
>      ├─ FixerDocumentGenerator (rule lookup)
>      └─ DescribeCommand (CLI display/expand)
> ```
>
> If normalization were only in `FixerFactory`, the other three consumers would receive raw FQCN keys and break. RuleSet is the right place precisely because it is the single source of truth that everything downstream reads from.
>
> The additional `normaliseRuleName()` call inside `FixerFactory::useRuleSet()` (line 175 in the diff) is a separate, defensive measure for custom `RuleSetInterface` implementations that might return un-normalized rules from getRules() — not a sign that normalization belongs there instead.

Wirone committed:
Fix typo in test cases

Wirone commented on src/FixerNameValidator.php:
```diff
@@ -29,6 +29,7 @@ public function isValid(string $name, bool $isCustom): bool
 
-        return Preg::match('/^[A-Z][a-zA-Z0-9]*\/[a-z][a-z0-9_]*$/', $name);
+        // See: https://regex101.com/r/UngJUn/6
+        return Preg::match('/^[a-z][a-z0-9]*(?!$)((?:\/(?!\d)[a-z0-9_]*)?|(?:(?<!PhpCsFixer)\\\(?!\d)[a-z0-9_]*)*)$/i', $name);
```
I believe [regex101](https://regex101.com/r/UngJUn/6) linked above the regex summarises it pretty well - it shows acceptable, and disallowed names.

Wirone commented:
@keradus I've rebased my branch with conflict resolution, during which:

- I dropped commit that changed to American english (it's now `RuleSet::normaliseConfig()` + `RuleSet::normaliseRuleName()`
- I added docs section about FQCN-based configuration

Besides that, I addressed your comments. I would like to know whether we're good to go now, or if there are still blockers. FYI: I would rather like to finish this one as-is, and then eventually create new PR with FQCN support for rule sets (including custom), because this PR has 3-year context and is really hard to comprehend at this point.

Wirone committed:
Increase code coverage in FixerFactory + fix bug

Sorting was invalid, it used key sorting, but that was already int-indexed array coming from `array_keys()`.

kubawerlos commented on tests/FixerNameValidatorTest.php:
```diff
@@ -80,12 +80,32 @@ public static function provideIsValidCases(): iterable
+
+        yield ['phpcsfixer\Restricted', true, false];
+
+        yield ['PHPCSFIXER\Restricted', true, false];
```
```suggestion
        yield ['PHPCSFIXER\Restricted', true, false];

        yield ['CustomFixersForPhpCsFixer\Allowed', true, true];
```

Wirone replied:
Addressed in [ff7eba0](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/7064/commits/ff7eba0c108123c5ac98cdb6edcf67f68072df89).

kubawerlos committed:
Ensure deprecation is reported for a fixer configured by its FQCN

kubawerlos commented on src/Documentation/RuleSetDocumentationGenerator.php:
```diff
@@ -104,7 +107,7 @@ static function (DocumentationTag $tag): string {
                         );
 
-                        $doc .= "\n- `{$rule} <{$path}>`_";
+                        $doc .= "\n- `{$fixerNames[$rule]->getName()} <{$path}>`_";
```
What is the reason for that change? Right now, it only introduces a new entry in PHPStan baseline. No test is failing without it.

If this is needed for something, please add a test.

Wirone replied:
Addressed in [b9bd347](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/7064/commits/b9bd3476f06ed4abd407c3d13cc5cbbedb56d30a).

kubawerlos commented on src/Console/ConfigurationResolver.php:
```diff
@@ -885,13 +883,13 @@ private function validateRules(array $rules): void
-        foreach ($fixers as $fixer) {
+        foreach ($this->createFixerFactory()->getFixers() as $fixer) {
             $fixerName = $fixer->getName();
             if (isset($rules[$fixerName]) && $fixer instanceof DeprecatedFixerInterface) {
```
Fixer configured by its FQCN will skip this check, right?

Let's have a test that will [ensure deprecation is reported for a fixer configured by its FQCN](https://github.com/Wirone/PHP-CS-Fixer/pull/13).

Wirone replied:
Addressed in 68e342e.

Wirone committed:
Allow using `PhpCsFixer` as part of fixer FQCN if it's not a leading part

Wirone committed:
Revert change in doc generator

At this point I don't remember why it was done like this, and since it works with previous version of the code, then let's use it (also it means one PHPStan issue less).
